### PR TITLE
issue 61346 route can be added and disposed in the same frame

### LINKED ISSUE #61346: Route throws null reference when TabBarView is combined with CupertinoTabView
When using `CupertinoTabView` inside a `TabBarView` the framework will throw `Unhandled Exception: NoSuchMethodError: The getter 'focusScopeNode' was called on null` when moving between tabs in a specific way.

We're using that combination of classes to get material design tabs but with iOS-style push navigation within the tab view.

This has been reported many times but without a repro:  #45539, #42829, #56827, and on Stack Overflow.

This is caused by [this line](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/navigator.dart#L216) in `Route`: the route gets disposed before the ticker completes, so `navigator` is null. 

## Steps to Reproduce

1. Run the sample below
2. Click on tab 3 **_without_** clicking on tab 2 first
3. Framework throws an exception:

```
Unhandled Exception: NoSuchMethodError: The getter 'focusScopeNode' was called on null.
Receiver: null
Tried calling: focusScopeNode
#0      Object.noSuchMethod  (dart:core-patch/object_patch.dart:53:5)
#1      Route.didAdd.<anonymous closure> 
package:flutter/…/widgets/navigator.dart:212
#2      _rootRunUnary  (dart:async/zone.dart:1192:38)
#3      _CustomZone.runUnary  (dart:async/zone.dart:1085:19)
#4      _FutureListener.handleValue  (dart:async/future_impl.dart:141:18)
#5      Future._propagateToListeners.handleValueCallback  (dart:async/future_impl.dart:682:45)
#6      Future._propagateToListeners  (dart:async/future_impl.dart:711:32)
#7      Future._completeWithValue  (dart:async/future_impl.dart:526:5)
#8      Future._asyncComplete.<anonymous closure>  (dart:async/future_impl.dart:556:7)
#9      _rootRun  (dart:async/zone.dart:1184:13)
#10     _CustomZone.run  (dart:async/zone.dart:1077:19)
#11     _CustomZone.runGuarded  (dart:async/zone.dart:979:7)
```
Note that this only happens when you "skip over" a tab by, in this case, going from the first to last.

**Expected results:** 

Tab changes, but an exception is thrown in the background.

**Actual results:** 

Tab changes without throwing an exception.

## Repro
```dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(home: HomePage());
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 3,
      child: Scaffold(
        appBar: AppBar(
          bottom: TabBar(
            tabs: <Widget>[
              Tab(text: "Tab 1"),
              Tab(text: "Tab 2"),
              Tab(text: "Tab 3"),
            ],
          ),
        ),
        body: TabBarView(
          children: <Widget>[
            CustomTabView(label: "Click tab 3 without clicking tab 2 first"),
            CustomTabView(label: "Tab 2"),
            CustomTabView(
              label:
                  "If you didn't click tab 2, you'll find an exception in your debug console",
            ),
          ],
        ),
      ),
    );
  }
}

class CustomTabView extends StatelessWidget {
  final String label;

  CustomTabView({this.label});

  @override
  Widget build(BuildContext context) {
    return CupertinoTabView(
      onGenerateRoute: (_) => CupertinoPageRoute(
        builder: (context) {
          return Center(child: Text(label));
        },
      ),
    );
  }
}
```

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.17.5 at /Users/tom/dev/flutter
    • Framework revision 8af6b2f038 (13 days ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/tom/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/tom/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (1 available)
    • iPad (7th generation) • 20451586-0B77-4E8D-9635-93F9A19159B9 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

The route may be added and disposed in the same frame and it throws error in such cases. This pr fixes it.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/61346

## Tests

I added the following tests:

see files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/navigator_test.dart [hunk 1]
```diff
@@ -489,6 +489,46 @@ void main() {
...
+        )
+      ),
+    );
+
+    dynamic error;
+
+    try {
+      controller.index = 2;
+      await tester.pumpAndSettle();
+    } catch (e){
```
Comment (goderbauer) [inline packages/flutter/test/widgets/navigator_test.dart:524]: nit: space before {

File: packages/flutter/test/widgets/navigator_test.dart [hunk 2]
```diff
@@ -489,6 +489,46 @@ void main() {
...
+      error = e;
+    }
+
+    expect(error, isNull);
+
```
Comment (goderbauer) [inline packages/flutter/test/widgets/navigator_test.dart:529]: nit: remove extra blank line

File: packages/flutter/test/widgets/navigator_test.dart [hunk 3]
```diff
@@ -489,6 +489,46 @@ void main() {
...
+    dynamic error;
+
+    try {
+      controller.index = 2;
+      await tester.pumpAndSettle();
+    } catch (e){
+      error = e;
+    }
+
+    expect(error, isNull);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/navigator_test.dart:528]: Do you need this? Wouldn't the test fail by itself if it throws an error?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 4]
```diff
@@ -213,7 +213,11 @@ abstract class Route<T> {
     // focused child can only be attached to navigator after initState which
     // will be guarded by the asynchronous gap.
     TickerFuture.complete().then<void>((void _) {
-      navigator.focusScopeNode.requestFocus();
+      // The route can be disposed before the ticker future completes. This can
+      // happen when the navigator is under a TabView that does a warping. The
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/navigator.dart:217]: does a warping => "warps" from one tab to another, non-adjacent tab, with an animation. 

On the face of it, it's not clear why this route might have been disposed as a consequence of the warp. The test appears to set up a situation where the old Navigator will have been disposed - not its route. Can you explain this scenario in a little more detail?

**Pull request merged**